README.md
# fluree-rpc-errors


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
(modification: no type change headlines) and this project adheres to
[Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [0.1.1] - 2020-09-26

### fixed

- json error json files add to dist

### Changed

- Documentation and test improvements

## [0.1.0] - 2020-09-25

- Basic functionality

[0.1.1]: https://github.com/StylusFrost/fluree-rpc-errors/compare/v0.1.0...v0.1.1
[0.1.0]: https://github.com/StylusFrost/fluree-rpc-errors/releases/v0.1.0


test/errors.spec.ts
/* tslint:disable no-invalid-this */
import * as assert from 'assert'

const dequal = require('fast-deep-equal')

import * as flureeErrors from '../src/errors'

import { FlureeRpcError, FlureeProviderError } from '../src/classes'

import { ERROR_CODES } from '../src/errors'

import { getMessageFromCode } from '../src/utils'

const rpcCodes = ERROR_CODES.rpc
const providerCodes = ERROR_CODES.provider
import { JSON_RPC_SERVER_ERROR_MESSAGE as serverErrorMessage } from '../src/utils'

const rpcCodeValues = Object.values(rpcCodes)

const dummyData = { foo: 'bar' }

const SERVER_ERROR_CODE = -32098
const CUSTOM_ERROR_CODE = 1001
const CUSTOM_ERROR_MESSAGE = 'foo'

describe('Errors test', () => {
  it('ensure exported object accepts a single string argument where appropriate', () => {
    let err = flureeErrors.rpc.invalidInput(CUSTOM_ERROR_MESSAGE)
    assert.strictEqual(err.code, ERROR_CODES.rpc.invalidInput, 'code is as expected')
    assert.strictEqual(err.message, CUSTOM_ERROR_MESSAGE, 'message is as expected')

    err = flureeErrors.provider.unauthorized(CUSTOM_ERROR_MESSAGE)
    assert.strictEqual(err.code, ERROR_CODES.provider.unauthorized, 'code is as expected')
    assert.strictEqual(err.message, CUSTOM_ERROR_MESSAGE, 'message is as expected')
  })

  it('custom provider error options', () => {
    assert.throws(
      function() {
        flureeErrors.provider.custom('bar')
      },
      /Fluree Provider custom errors must/u,
      'does not accept non-object argument',
    )

    assert.throws(
      function() {
        flureeErrors.provider.custom({ code: 4009, message: 2 })
      },
      /"message" must be/u,
      'does not accept nonstring message',
    )

    assert.throws(
      function() {
        flureeErrors.provider.custom({ code: 4009, message: '' })
      },
      /"message" must be/u,
      'does not accept empty message',
    )

    flureeErrors.provider.custom({ code: 4009, message: 'foo' })
  })

  it('server rpc error options', () => {
    assert.throws(
      function() {
        flureeErrors.rpc.server('bar')
      },
      /Fluree RPC Server errors must/u,
      'does not accept non-object argument',
    )

    assert.throws(
      function() {
        flureeErrors.rpc.server({ code: 'bar' })
      },
      /"code" must be/u,
      'does not accept non-string code',
    )

    assert.throws(
      function() {
        flureeErrors.rpc.server({ code: 1 })
      },
      /"code" must be/u,
      'does not accept out-of-range code',
    )

    flureeErrors.rpc.server({ code: -32008, message: 'foo' })
  })

  // we just iterate over all keys on the errors object and call and check
  // each error in turn
  it('test exported object for correctness', () => {
    it('Begin: Fluree RPC', () =>
      Object.keys(flureeErrors.rpc).forEach(k => {
        if (k === 'server') {
          validateError(
            flureeErrors.rpc[k]({
              code: SERVER_ERROR_CODE,
              message: null,
              data: Object.assign({}, dummyData),
            }),
            k,
            dummyData,
          )
        } else {
          validateError(
            flureeErrors.rpc[k]({
              message: null,
              data: Object.assign({}, dummyData),
            }),
            k,
            dummyData,
          )
        }
      }))

    it('Begin: Fluree Provider', () =>
      Object.keys(flureeErrors.provider).forEach(k => {
        if (k === 'custom') {
          validateError(
            flureeErrors.provider[k]({
              code: CUSTOM_ERROR_CODE,
              message: CUSTOM_ERROR_MESSAGE,
              data: Object.assign({}, dummyData),
            }),
            k,
            dummyData,
            true,
          )
        } else {
          validateError(
            flureeErrors.provider[k]({
              message: null,
              data: Object.assign({}, dummyData),
            }),
            k,
            dummyData,
            true,
          )
        }
      }))
  })

  function validateError(err: any, key: string, data: any, isProvider: Boolean = false) {
    assert.ok(
      err instanceof Error && err instanceof FlureeRpcError && isProvider
        ? err instanceof FlureeProviderError
        : true,
      'error has correct inheritance',
    )
    assert.ok(Number.isInteger(err.code), 'code is an integer')
    assert.ok(err.message && typeof err.message === 'string', 'message is a string')
    assert.ok(dequal(err.data, data), 'data is as provided')

    if (err instanceof FlureeProviderError) {
      assert.ok(err.code >= 1000 && err.code < 5000, 'code has valid value')

      if (key === 'custom') {
        assert.ok(
          err.code === CUSTOM_ERROR_CODE && err.message === CUSTOM_ERROR_MESSAGE,
          'code and message values correspond for error type',
        )
      } else {
        assert.ok(
          err.code === providerCodes[key] && err.message === getMessageFromCode(providerCodes[key]),
          'code and message values correspond for error type',
        )
      }
    } else if (err instanceof FlureeRpcError) {
      assert.ok(
        rpcCodeValues.includes(err.code) || (err.code <= -32000 && err.code >= -32099),
        'code has valid value',
      )

      if (key === 'server') {
        assert.ok(
          err.code <= -32000 && err.code >= -32099 && err.message === serverErrorMessage,
          'code and message values correspond for error type',
        )
      } else {
        assert.ok(
          err.code === rpcCodes[key] && err.message === getMessageFromCode(rpcCodes[key]),
          'code and message values correspond for error type',
        )
      }
    }
  }
})


test/serializeErrors.spec.ts
import { assert } from 'console'

const dequal = require('fast-deep-equal')

import { flureeErrors } from '../src/errors'

import { ERROR_CODES } from '../src/errors'

import { getMessageFromCode, serializeError } from '../src/utils'
const rpcCodes = ERROR_CODES.rpc

const dummyData = { foo: 'bar' }
const dummyMessage = 'baz'

const invalidError0 = 0
const invalidError1 = ['foo', 'bar', 3]
const invalidError2 = { code: 34 }
const invalidError3 = { code: 4001 }
const invalidError4 = { code: 4001, message: 3, data: Object.assign({}, dummyData) }
const invalidError5 = null
const invalidError6 = undefined
const invalidError7 = { code: 34, message: dummyMessage, data: Object.assign({}, dummyData) }

const validError0 = { code: 4001, message: dummyMessage }
const validError1 = { code: 4001, message: dummyMessage, data: Object.assign({}, dummyData) }
const validError2 = flureeErrors.rpc.parse()
delete validError2.stack
const validError3 = flureeErrors.rpc.parse(dummyMessage)
delete validError3.stack
const validError4 = flureeErrors.rpc.parse({
  message: dummyMessage,
  data: Object.assign({}, dummyData),
})
delete validError4.stack

describe('Serialize Errors test', () => {
  it('invalid error: non-object', () => {
    const result = serializeError(invalidError0)
    assert(
      dequal(result, {
        code: rpcCodes.internal,
        message: getMessageFromCode(rpcCodes.internal),
        data: { originalError: invalidError0 },
      }),
      'serialized error matches expected result',
    )
  })

  it('invalid error: null', () => {
    const result = serializeError(invalidError5)
    assert(
      dequal(result, {
        code: rpcCodes.internal,
        message: getMessageFromCode(rpcCodes.internal),
        data: { originalError: invalidError5 },
      }),
      'serialized error matches expected result',
    )
  })

  it('invalid error: undefined', () => {
    const result = serializeError(invalidError6)
    assert(
      dequal(result, {
        code: rpcCodes.internal,
        message: getMessageFromCode(rpcCodes.internal),
        data: { originalError: invalidError6 },
      }),
      'serialized error matches expected result',
    )
  })

  it('invalid error: array', () => {
    const result = serializeError(invalidError1)
    assert(
      dequal(result, {
        code: rpcCodes.internal,
        message: getMessageFromCode(rpcCodes.internal),
        data: { originalError: invalidError1 },
      }),
      'serialized error matches expected result',
    )
  })

  it('invalid error: invalid code', () => {
    const result = serializeError(invalidError2)
    assert(
      dequal(result, {
        code: rpcCodes.internal,
        message: getMessageFromCode(rpcCodes.internal),
        data: { originalError: Object.assign({}, invalidError2) },
      }),
      'serialized error matches expected result',
    )
  })

  it('invalid error: valid code, undefined message', () => {
    const result = serializeError(invalidError3)
    assert(
      dequal(result, {
        code: 4001,
        message: getMessageFromCode(4001),
        data: { originalError: Object.assign({}, invalidError3) },
      }),
      'serialized error matches expected result',
    )
  })

  it('invalid error: non-string message with data', () => {
    const result = serializeError(invalidError4)
    assert(
      dequal(result, {
        code: 4001,
        message: getMessageFromCode(4001),
        data: { originalError: Object.assign({}, invalidError4) },
      }),
      'serialized error matches expected result',
    )
  })

  it('invalid error: invalid code with string message', () => {
    const result = serializeError(invalidError7)
    assert(
      dequal(result, {
        code: rpcCodes.internal,
        message: dummyMessage,
        data: { originalError: Object.assign({}, invalidError7) },
      }),
      'serialized error matches expected result',
    )
  })

  it('invalid error: invalid code, no message, custom fallback', () => {
    const result = serializeError(invalidError2, {
      fallbackError: { code: rpcCodes.methodNotFound, message: 'foo' },
    })
    assert(
      dequal(result, {
        code: rpcCodes.methodNotFound,
        message: 'foo',
        data: { originalError: Object.assign({}, invalidError2) },
      }),
      'serialized error matches expected result',
    )
  })

  it('valid error: code and message only', () => {
    const result = serializeError(validError0)
    assert(
      dequal(result, {
        code: 4001,
        message: validError0.message,
      }),
      'serialized error matches expected result',
    )
  })

  it('valid error: code, message, and data', () => {
    const result = serializeError(validError1)
    assert(
      dequal(result, {
        code: 4001,
        message: validError1.message,
        data: Object.assign({}, validError1.data),
      }),
      'serialized error matches expected result',
    )
  })

  it('valid error: instantiated error', () => {
    const result = serializeError(validError2)
    assert(
      dequal(result, {
        code: rpcCodes.parse,
        message: getMessageFromCode(rpcCodes.parse),
      }),
      'serialized error matches expected result',
    )
  })

  it('valid error: instantiated error', () => {
    const result = serializeError(validError3)
    assert(
      dequal(result, {
        code: rpcCodes.parse,
        message: dummyMessage,
      }),
      'serialized error matches expected result',
    )
  })

  it('valid error: instantiated error with custom message and data', () => {
    const result = serializeError(validError4)
    assert(
      dequal(result, {
        code: rpcCodes.parse,
        message: validError4.message,
        data: Object.assign({}, validError4.data),
      }),
      'serialized error matches expected result',
    )
  })

  it('valid error: message, data, and stack', () => {
    const result = serializeError(Object.assign({}, validError1, { stack: 'foo' }))
    assert(
      dequal(result, {
        code: 4001,
        message: validError1.message,
        data: Object.assign({}, validError1.data),
      }),
      'serialized error matches expected result',
    )
  })

  it('include stack: no stack present', () => {
    const result = serializeError(validError1, { shouldIncludeStack: true })
    assert(
      dequal(result, {
        code: 4001,
        message: validError1.message,
        data: Object.assign({}, validError1.data),
      }),
      'serialized error matches expected result',
    )
  })

  it('include stack: string stack present', () => {
    const result = serializeError(Object.assign({}, validError1, { stack: 'foo' }), {
      shouldIncludeStack: true,
    })
    assert(
      dequal(result, {
        code: 4001,
        message: validError1.message,
        data: Object.assign({}, validError1.data),
        stack: 'foo',
      }),
      'serialized error matches expected result',
    )
  })

  it('include stack: non-string stack present', () => {
    const result = serializeError(Object.assign({}, validError1, { stack: 2 }), {
      shouldIncludeStack: true,
    })
    assert(
      dequal(result, {
        code: 4001,
        message: validError1.message,
        data: Object.assign({}, validError1.data),
      }),
      'serialized error matches expected result',
    )
  })
})


src/utils.ts
const errorValues = require('./errorValues')
const FALLBACK_ERROR_CODE = require('./errorCodes.json').rpc.internal
import { FlureeRpcError } from './classes'

export const JSON_RPC_SERVER_ERROR_MESSAGE = 'Unspecified server error.'

const FALLBACK_MESSAGE = 'Unspecified error message. This is a bug, please report it.'

const FALLBACK_ERROR = {
  code: FALLBACK_ERROR_CODE,
  message: getMessageFromCode(FALLBACK_ERROR_CODE),
}

/**
 * Gets the message for a given code, or a fallback message if the code has
 * no corresponding message.
 *
 * @param {number} code - The integer error code
 * @param {string} fallbackMessage - The fallback message
 * @return {string} The corresponding message or the fallback message
 */
export function getMessageFromCode(code: number, fallbackMessage = FALLBACK_MESSAGE) {
  if (Number.isInteger(code)) {
    const codeString = code.toString()

    if (errorValues[codeString]) {
      return errorValues[codeString].message
    }
    if (isJsonRpcServerError(code)) {
      return JSON_RPC_SERVER_ERROR_MESSAGE
    }
  }
  return fallbackMessage
}

/**
 * Returns whether the given code is valid.
 * A code is only valid if it has a message.
 *
 * @param {Number} code - The code to check
 * @return {boolean} true if the code is valid, false otherwise.
 */
export function isValidCode(code: number): boolean {
  if (!Number.isInteger(code)) {
    return false
  }

  const codeString = code.toString()
  if (errorValues[codeString]) {
    return true
  }

  if (isJsonRpcServerError(code)) {
    return true
  }

  return false
}

/**
 * Serializes the given error to an Fluree JSON RPC-compatible error object.
 * Merely copies the given error's values if it is already compatible.
 * If the given error is not fully compatible, it will be preserved on the
 * returned object's data.originalError property.
 *
 * @param {any} error - The error to serialize.
 * @param {Object} [options] - An options object.
 * @param {Object} [options.fallbackError] - The custom fallback error values if
 * the given error is invalid.
 * @param {boolean} [options.shouldIncludeStack] - Whether the 'stack' property
 * of the given error should be included on the serialized error, if present.
 * @return {Object} A standardized, plain error object.
 */
export function serializeError(
  error: any,
  { fallbackError = FALLBACK_ERROR, shouldIncludeStack = false } = {},
) {
  if (
    !fallbackError ||
    !Number.isInteger(fallbackError.code) ||
    typeof fallbackError.message !== 'string'
  ) {
    throw new Error('Must provide fallback error with integer number code and string message.')
  }

  if (error instanceof FlureeRpcError) {
    return error.serialize()
  }

  const serialized: any = {}

  if (error && isValidCode(error.code)) {
    serialized['code'] = error.code

    if (error.message && typeof error.message === 'string') {
      serialized['message'] = error.message
      if ('data' in error) {
        serialized['data'] = error.data
      }
    } else {
      serialized['message'] = getMessageFromCode(serialized['code'])
      serialized['data'] = { originalError: assignOriginalError(error) }
    }
  } else {
    serialized['code'] = fallbackError['code']
    serialized['message'] = error && error.message ? error.message : fallbackError.message
    serialized['data'] = { originalError: assignOriginalError(error) }
  }

  if (shouldIncludeStack && error && typeof error.stack === 'string') {
    serialized['stack'] = error.stack
  }
  return serialized
}

// Internal

function isJsonRpcServerError(code: Number) {
  return code >= -32099 && code <= -32000
}

function assignOriginalError(error: any) {
  if (error && typeof error === 'object' && !Array.isArray(error)) {
    return Object.assign({}, error)
  }
  return error
}


src/classes.ts
const safeStringify = require('fast-safe-stringify')

/**
 * @class JsonRpcError
 * Error subclass implementing JSON RPC 2.0 errors and Fluree RPC errors
 * Permits any integer error code.
 */
export class FlureeRpcError extends Error {
  code: Number
  data: any

  /**
   * Create an Fluree JSON RPC error.
   *
   * @param {Number} code - The integer error code.
   * @param {string} message - The string message.
   * @param {any} [data] - The error data.
   */
  constructor(code: number, message: string, data: any) {
    super(message)
    if (!Number.isInteger(code)) {
      throw new Error('"code" must be an integer.')
    }
    if (!message || typeof message !== 'string') {
      throw new Error('"message" must be a nonempty string.')
    }

    this.code = code
    if (data !== undefined) {
      this.data = data
    }
  }

  /**
   * Returns a plain object with all public class properties.
   *
   * @returns {object} The serialized error.
   */
  public serialize() {
    const serialized: any = {
      code: this.code,
      message: this.message,
    }
    if (this.data !== undefined) {
      serialized['data'] = this.data
    }
    if (this.stack) {
      serialized['stack'] = this.stack
    }
    return serialized
  }

  /**
   * Return a string representation of the serialized error, omitting
   * any circular references.
   *
   * @returns {string} The serialized error as a string.
   */
  toString() {
    return safeStringify(this.serialize(), stringifyReplacer, 2)
  }
}

/**
 * @class FlureeRpcError
 * Error subclass implementing Fluree Provider errors .
 * Permits integer error codes in the [ 1000 <= 4999 ] range.
 */
export class FlureeProviderError extends FlureeRpcError {
  /**
   * Create an Fluree JSON RPC error.
   *
   * @param {number} code - The integer error code, in the [ 1000 <= 4999 ] range.
   * @param {string} message - The string message.
   * @param {any} [data] - The error data.
   */
  constructor(code: number, message: string, data: any) {
    super(code, message, data)

    if (!isValidFlureeProviderCode(code)) {
      throw new Error('"code" must be an integer such that: 1000 <= code <= 4999')
    }
  }
}

// Internal

function isValidFlureeProviderCode(code: number) {
  return Number.isInteger(code) && code >= 1000 && code <= 4999
}

function stringifyReplacer(_: any, value: string) {
  if (value === '[Circular]') {
    return undefined
  }
  return value
}


src/errors.ts
import { FlureeRpcError, FlureeProviderError } from './classes'
import { getMessageFromCode } from './utils'
export const ERROR_CODES = require('./errorCodes.json')

export const rpc: any = {
  /**
   * Get a JSON RPC 2.0 Parse (-32700) error.
   *
   * @param {Object|string} [opts] - Options object or error message string
   * @param {string} [opts.message] - The error message
   * @param {any} [opts.data] - Error data
   * @returns {FlureeRpcError} The error
   */
  parse: (opts: Object | string) => getFlureeJsonRpcError(ERROR_CODES.rpc.parse, opts),

  /**
   * Get a JSON RPC 2.0 Invalid Request (-32600) error.
   *
   * @param {Object|string} [opts] - Options object or error message string
   * @param {string} [opts.message] - The error message
   * @param {any} [opts.data] - Error data
   * @returns {FlureeRpcError} The error
   */
  invalidRequest: (opts: Object | string) =>
    getFlureeJsonRpcError(ERROR_CODES.rpc.invalidRequest, opts),

  /**
   * Get a JSON RPC 2.0 Invalid Params (-32602) error.
   *
   * @param {Object|string} [opts] - Options object or error message string
   * @param {string} [opts.message] - The error message
   * @param {any} [opts.data] - Error data
   * @returns {FlureeRpcError} The error
   */
  invalidParams: (opts: Object | string) =>
    getFlureeJsonRpcError(ERROR_CODES.rpc.invalidParams, opts),

  /**
   * Get a JSON RPC 2.0 Method Not Found (-32601) error.
   *
   * @param {Object|string} [opts] - Options object or error message string
   * @param {string} [opts.message] - The error message
   * @param {any} [opts.data] - Error data
   * @returns {FlureeRpcError} The error
   */
  methodNotFound: (opts: Object | string) =>
    getFlureeJsonRpcError(ERROR_CODES.rpc.methodNotFound, opts),

  /**
   * Get a JSON RPC 2.0 Internal (-32603) error.
   *
   * @param {Object|string} [opts] - Options object or error message string
   * @param {string} [opts.message] - The error message
   * @param {any} [opts.data] - Error data
   * @returns {FlureeRpcError} The error
   */
  internal: (opts: Object | string) => getFlureeJsonRpcError(ERROR_CODES.rpc.internal, opts),

  /**
   * Get a JSON RPC 2.0 Server error.
   * Permits integer error codes in the [ -32099 <= -32008 ] range.
   * Codes -32000 through -32007 are reserved
   *
   * @param {Object|string} opts - Options object
   * @param {number} opts.code - The error code
   * @param {string} [opts.message] - The error message
   * @param {any} [opts.data] - Error data
   * @returns {FlureeRpcError} The error
   */
  server: (opts: any) => {
    if (!opts || typeof opts !== 'object' || Array.isArray(opts)) {
      throw new Error('Fluree RPC Server errors must provide single object argument.')
    }
    const { code } = opts
    if (!Number.isInteger(code) || code > -32007 || code < -32099) {
      throw new Error('"code" must be an integer such that: -32099 <= code <= -32008')
    }
    return getFlureeJsonRpcError(code, opts)
  },

  /**
   * Get an Fluree JSON RPC Invalid Input (-32000) error.
   *
   * @param {Object|string} [opts] - Options object or error message string
   * @param {string} [opts.message] - The error message
   * @param {any} [opts.data] - Error data
   * @returns {FlureeRpcError} The error
   */
  invalidInput: (opts: Object | string) =>
    getFlureeJsonRpcError(ERROR_CODES.rpc.invalidInput, opts),

  /**
   * Get an Fluree JSON RPC Resource Not Found (-32001) error.
   *
   * @param {Object|string} [opts] - Options object or error message string
   * @param {string} [opts.message] - The error message
   * @param {any} [opts.data] - Error data
   * @returns {FlureeRpcError} The error
   */
  resourceNotFound: (opts: Object | string) =>
    getFlureeJsonRpcError(ERROR_CODES.rpc.resourceNotFound, opts),

  /**
   * Get an Fluree JSON RPC Resource Unavailable (-32002) error.
   *
   * @param {Object|string} [opts] - Options object or error message string
   * @param {string} [opts.message] - The error message
   * @param {any} [opts.data] - Error data
   * @returns {FlureeRpcError} The error
   */
  resourceUnavailable: (opts: Object | string) =>
    getFlureeJsonRpcError(ERROR_CODES.rpc.resourceUnavailable, opts),

  /**
   * Get an Fluree JSON RPC Transaction Rejected (-32003) error.
   *
   * @param {Object|string} [opts] - Options object or error message string
   * @param {string} [opts.message] - The error message
   * @param {any} [opts.data] - Error data
   * @returns {FlureeRpcError} The error
   */
  transactionRejected: (opts: Object | string) =>
    getFlureeJsonRpcError(ERROR_CODES.rpc.transactionRejected, opts),

  /**
   * Get an Fluree JSON RPC Query Rejected (-32004) error.
   *
   * @param {Object|string} [opts] - Options object or error message string
   * @param {string} [opts.message] - The error message
   * @param {any} [opts.data] - Error data
   * @returns {FlureeRpcError} The error
   */
  queryRejected: (opts: Object | string) =>
    getFlureeJsonRpcError(ERROR_CODES.rpc.queryRejected, opts),

  /**
   * Get an Fluree JSON RPC Request Rejected (-32005) error.
   *
   * @param {Object|string} [opts] - Options object or error message string
   * @param {string} [opts.message] - The error message
   * @param {any} [opts.data] - Error data
   * @returns {FlureeRpcError} The error
   */
  requestRejected: (opts: Object | string) =>
    getFlureeJsonRpcError(ERROR_CODES.rpc.requestRejected, opts),

  /**
   * Get an Fluree JSON RPC Method Not Supported (-32006) error.
   *
   * @param {Object|string} [opts] - Options object or error message string
   * @param {string} [opts.message] - The error message
   * @param {any} [opts.data] - Error data
   * @returns {FlureeRpcError} The error
   */
  methodNotSupported: (opts: Object | string) =>
    getFlureeJsonRpcError(ERROR_CODES.rpc.methodNotSupported, opts),

  /**
   * Get an Fluree JSON RPC Limit Exceeded (-32007) error.
   *
   * @param {Object|string} [opts] - Options object or error message string
   * @param {string} [opts.message] - The error message
   * @param {any} [opts.data] - Error data
   * @returns {FlureeRpcError} The error
   */
  limitExceeded: (opts: Object | string) =>
    getFlureeJsonRpcError(ERROR_CODES.rpc.limitExceeded, opts),
}

export const provider: any = {
  /**
   * Get an Fluree Provider User Rejected Request (4001) error.
   *
   * @param {Object|string} [opts] - Options object or error message string
   * @param {string} [opts.message] - The error message
   * @param {any} [opts.data] - Error data
   * @returns {FlureeProviderError} The error
   */
  userRejectedRequest: (opts: Object | string) => {
    return getFlureeProviderError(ERROR_CODES.provider.userRejectedRequest, opts)
  },

  /**
   * Get an Fluree Provider Unauthorized (4100) error.
   *
   * @param {Object|string} [opts] - Options object or error message string
   * @param {string} [opts.message] - The error message
   * @param {any} [opts.data] - Error data
   * @returns {FlureeProviderError} The error
   */
  unauthorized: (opts: Object | string) => {
    return getFlureeProviderError(ERROR_CODES.provider.unauthorized, opts)
  },

  /**
   * Get an Fluree Provider Unsupported Method (4200) error.
   *
   * @param {Object|string} [opts] - Options object or error message string
   * @param {string} [opts.message] - The error message
   * @param {any} [opts.data] - Error data
   * @returns {FlureeProviderError} The error
   */
  unsupportedMethod: (opts: Object | string) => {
    return getFlureeProviderError(ERROR_CODES.provider.unsupportedMethod, opts)
  },

  /**
   * Get an Fluree Provider Not Connected (4900) error.
   *
   * @param {Object|string} [opts] - Options object or error message string
   * @param {string} [opts.message] - The error message
   * @param {any} [opts.data] - Error data
   * @returns {FlureeProviderError} The error
   */
  disconnected: (opts: Object | string) => {
    return getFlureeProviderError(ERROR_CODES.provider.disconnected, opts)
  },

  /**
   * Get an Fluree Provider Database Not Connected (4901) error.
   *
   * @param {Object|string} [opts] - Options object or error message string
   * @param {string} [opts.message] - The error message
   * @param {any} [opts.data] - Error data
   * @returns {FlureeProviderError} The error
   */
  databaseDisconnected: (opts: Object | string) => {
    return getFlureeProviderError(ERROR_CODES.provider.databaseDisconnected, opts)
  },

  /**
   * Get a custom Fluree Provider error.
   *
   * @param {Object|string} opts - Options object
   * @param {number} opts.code - The error code
   * @param {string} opts.message - The error message
   * @param {any} [opts.data] - Error data
   * @returns {FlureeProviderError} The error
   */
  custom: (opts: any) => {
    if (!opts || typeof opts !== 'object' || Array.isArray(opts)) {
      throw new Error('Fluree Provider custom errors must provide single object argument.')
    }
    const { code, message, data } = opts
    if (!message || typeof message !== 'string') {
      throw new Error('"message" must be a nonempty string')
    }
    return new FlureeProviderError(code, message, data)
  },
}

// Internal

function getFlureeJsonRpcError(code: number, opts: Object | string) {
  const [message, data] = validateOpts(opts)
  return new FlureeRpcError(code, message || getMessageFromCode(code), data)
}

function getFlureeProviderError(code: number, opts: Object | string) {
  const [message, data] = validateOpts(opts)
  return new FlureeProviderError(code, message || getMessageFromCode(code), data)
}

function validateOpts(opts: any) {
  if (opts) {
    if (typeof opts === 'string') {
      return [opts]
    } else if (typeof opts === 'object' && !Array.isArray(opts)) {
      const { message, data } = opts
      return [message, data]
    }
  }
  return []
}

export const flureeErrors: any = {
  rpc: rpc,
  provider: provider,
}
